### PR TITLE
Remove mono specific SpanHelpers

### PR DESCRIPTION
**Pull request by BrzVlad**
#73768 did various changes that hurt span performance on mono. The change was reverted afterwards on mono by restoring old code in #75917.

This PR removes mono specific code and solves performance problems via small tweaks within non vectorized code inside SpanHelpers, as well as by adding a couple of optimizations to mono interpreter.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs [hunk 1]
```diff
@@ -2075,7 +2075,6 @@ public static int SequenceCompareTo<T>(this Span<T> span, ReadOnlySpan<T> other)
                                 Unsafe.Add(ref valueRef, 2),
                                 span.Length);
 
-#if !MONO // We don't have a mono overload for 4 values
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs:2078]: IndexOfAny4Values.cs has one more MONO ifdef that should be reverted.

**Comment by BrzVlad** [comment]
@vargaz Could you take another look at this ?

**Review by vargaz — APPROVED** [review]

**Comment by radekdoulik** [comment]
Nice! I see interpreter improvements in the SpanHelper bytes measurements. The chars are mixed, improvements in firefox, some slower in chrome. Overall it is great improvement.
image
https://radekdoulik.github.io/WasmPerformanceMeasurements/?startDate=2022-12-05T10%3A43%3A18.830Z&endDate=2022-12-19T10%3A43%3A18.830Z&tasks=&flavors=2%2C3

It also improved Index of chars with the aot/simd. I think that's because the current code uses `S.R.I.Vector128`, where we have better coverage, compared to the older, which used `S.N.Vector`.
image
https://radekdoulik.github.io/WasmPerformanceMeasurements/?startDate=2022-12-05T10%3A43%3A18.830Z&endDate=2022-12-19T10%3A43%3A18.830Z&tasks=&flavors=4%2C8

/cc @lewing

**Comment by MihaZupan** [comment]
@BrzVlad can you please take a look at the changes in #78861 (they only affect platforms with Sse2/AdvSimd support). Should those be updated to better match what was done in this PR? I'm trying to avoid accidentally undoing the improvements you made here.

**Comment by BrzVlad** [comment]
@MihaZupan I don't think there are any problems between these 2 PRs. As long as your code is guarded everywhere with `IsSupported` / `IsHardwareAccelerated` then there should be changed for mono interpreter.

**Pull request merged**